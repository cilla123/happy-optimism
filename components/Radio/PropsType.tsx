export interface BaseRadioProps {
  theme?: 'default' | 'primary' | 'success' | 'warning' | 'error';
  shape?: 'radius' | 'round';
  type?: 'button' | 'cell';
  disabled?: boolean;
  block?: boolean;
  onChange?: (checked: boolean) => void;
  checked?: boolean;
  defaultChecked?: boolean;
  value?: number | string;
}

export interface BaseRadioGroupProps {
  theme?: 'default' | 'primary' | 'success' | 'warning' | 'error';
  shape?: 'radius' | 'round';
  size?: 'lg';
  type?: 'button' | 'cell';
  disabled?: boolean;
  block?: boolean;
  onChange?: (value: string | number) => void;
  value?: number | string;
  defaultValue?: number | string;
  compact?: boolean;
}
