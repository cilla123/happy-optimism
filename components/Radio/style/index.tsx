import '../../style/';
import '../../Cell/style/';
import '../../Button/style/';
import '../../Icon/style/';
import './index.scss';
