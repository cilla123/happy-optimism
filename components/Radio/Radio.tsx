import React, { PureComponent } from 'react';
import classnames from 'classnames';
import { BaseRadioProps } from './PropsType';
import Cell from '../Cell';
import Button from '../Button';
import Icon from '../Icon';

const getChecked = (props, defaultChecked) => {
  if ('checked' in props && props.checked) {
    return props.checked;
  }
  if ('defaultChecked' in props && props.defaultChecked) {
    return props.defaultChecked;
  }
  return defaultChecked;
};

export interface RadioProps extends BaseRadioProps {
  prefixCls?: string;
  className?: string;
}

export default class Radio extends PureComponent<RadioProps, any> {

  static Group: any;
  static defaultProps = {
    prefixCls: 'za-radio',
    theme: 'primary',
    disabled: false,
    block: false,
  };

  constructor(props) {
    super(props);
    this.state = {
      checked: getChecked(props, false),
    };
  }

  componentWillReceiveProps(nextProps) {
    if ('checked' in nextProps) {
      this.setState({
        checked: !!nextProps.checked,
      });
    }
  }

  onValueChange = () => {
    const { disabled, onChange } = this.props;

    if (disabled) {
      return;
    }

    const checked = true;
    this.setState({ checked });
    if (typeof onChange === 'function') {
      onChange(checked);
    }
  }

  render() {
    const { prefixCls, className, type, theme, shape, block, value, disabled, children } = this.props;
    const { checked } = this.state;

    const cls = classnames(`${prefixCls}`, className, {
      [`theme-${theme}`]: !!theme,
      [`shape-${shape}`]: !!shape,
      checked,
      disabled,
    });

    if (type === 'cell') {
      return (
        <Cell
          disabled={disabled}
          description={checked && <Icon type="right" theme={disabled ? undefined : theme} />}
          onClick={() => {}}
        >
          <input
            type="radio"
            className={`${prefixCls}-input`}
            value={value}
            disabled={disabled}
            checked={checked}
            onChange={this.onValueChange}
          />
          {children}
        </Cell>
      );
    }

    if (type === 'button') {
      return (
        <Button
          className={cls}
          theme={theme}
          shape={shape}
          size="sm"
          block={block}
          bordered={!checked}
          disabled={disabled}
        >
          <input
            type="radio"
            className={`${prefixCls}-input`}
            value={value}
            disabled={disabled}
            checked={checked}
            onChange={this.onValueChange}
          />
          {children}
        </Button>
      );
    }

    return (
      <div className={cls}>
        <div className={`${prefixCls}-wrapper`}>
          <span className={`${prefixCls}-inner`} />
          {children && <span className={`${prefixCls}-text`}>{children}</span>}
          <input
            type="radio"
            className={`${prefixCls}-input`}
            disabled={disabled}
            checked={checked}
            onChange={this.onValueChange}
          />
        </div>
      </div>
    );
  }
}
