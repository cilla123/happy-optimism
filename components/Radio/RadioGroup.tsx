import React, { PureComponent, cloneElement } from 'react';
import classnames from 'classnames';
import { BaseRadioGroupProps } from './PropsType';

const getChildChecked = (children) => {
  let checkedValue = null;
  React.Children.forEach(children, (element: any) => {
    if (element.props && element.props.checked) {
      checkedValue = element.props.value;
    }
  });
  return checkedValue;
};

const getValue = (props, defaultValue) => {
  if ('value' in props) {
    return props.value;
  }
  if ('defaultValue' in props) {
    return props.defaultValue;
  }
  if (getChildChecked(props.children)) {
    return getChildChecked(props.children);
  }
  return defaultValue;
};

export interface RadioGroupProps extends BaseRadioGroupProps {
  prefixCls?: string;
  className?: string;
}

export default class RadioGroup extends PureComponent<RadioGroupProps, any> {

  static defaultProps = {
    prefixCls: 'za-radio-group',
    theme: 'primary',
    block: false,
    disabled: false,
    compact: false,
  };

  constructor(props) {
    super(props);
    this.state = {
      value: getValue(props, null),
    };
  }

  componentWillReceiveProps(nextProps) {
    if ('value' in nextProps || getChildChecked(nextProps.children)) {
      this.setState({
        value: nextProps.value || getChildChecked(nextProps.children),
      });
    }
  }

  onChildChange = (value) => {
    this.setState({ value });
    const { onChange } = this.props;
    if (typeof onChange === 'function') {
      onChange(value);
    }
  }

  render() {
    const { prefixCls, className, theme, shape, type, block, disabled, compact, children } = this.props;

    const items = React.Children.map(children, (element: any, index) => {
      return cloneElement(element, {
        key: index,
        type,
        theme,
        shape,
        block: block || element.props.block,
        disabled: disabled || element.props.disabled,
        onChange: () => this.onChildChange(element.props.value),
        checked: this.state.value === element.props.value || Number(this.state.value) === Number(element.props.value),
      });
    });

    const cls = classnames(`${prefixCls}`, className, {
      [`shape-${shape}`]: !!shape,
      'is-compact': compact,
      block,
      disabled,
    });

    return <div className={cls}>{items}</div>;
  }
}
