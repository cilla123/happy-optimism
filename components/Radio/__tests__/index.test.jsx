import React from 'react';
import { render, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Radio from '../index';

describe('Radio', () => {
  it('renders correctly', () => {
    const wrapper = render(
      <Radio value="0">选项一</Radio>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('defaultChecked', () => {
    const wrapper = render(
      <Radio defaultChecked value="0">选项一</Radio>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('type is button', () => {
    const wrapper = render(
      <Radio type="button" value="0">选项一</Radio>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('type is cell', () => {
    const wrapper = render(
      <Radio type="cell" value="0">选项一</Radio>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('receive new checked', () => {
    const onChange = jest.fn();
    const wrapper = shallow(
      <Radio value="0" onChange={onChange}>选项一</Radio>
    );
    wrapper.setProps({ checked: true });
    wrapper.find('input[type="radio"]').simulate('change');
    expect(onChange).toBeCalled();
  });

  it('disabled', () => {
    const onChange = jest.fn();
    const wrapper = shallow(
      <Radio value="0" onChange={onChange}>选项一</Radio>
    );
    wrapper.setProps({ disabled: true });
    wrapper.find('input[type="radio"]').simulate('change');
  });
});

describe('Radio.Group', () => {
  it('renders correctly', () => {
    const wrapper = render(
      <Radio.Group value="0" onChange={jest.fn()}>
        <Radio value="0">选项一</Radio>
        <Radio value="1">选项二</Radio>
        <Radio value="2">选项三</Radio>
      </Radio.Group>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('receive new value', () => {
    const wrapper = shallow(
      <Radio.Group value="0" onChange={jest.fn()}>
        <Radio value="0">选项一</Radio>
        <Radio value="1">选项二</Radio>
        <Radio value="2">选项三</Radio>
      </Radio.Group>
    );
    wrapper.setProps({ value: '1' });
  });

  it('defaultValue', () => {
    const wrapper = render(
      <Radio.Group defaultValue="1">
        <Radio value="0">选项一</Radio>
        <Radio value="1">选项二</Radio>
        <Radio value="2">选项三</Radio>
      </Radio.Group>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('Radio checked', () => {
    const wrapper = render(
      <Radio.Group>
        <Radio value="0">选项一</Radio>
        <Radio value="1" checked>选项二</Radio>
        <Radio value="2">选项三</Radio>
      </Radio.Group>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  // 圆角
  it('shape is radius', () => {
    const wrapper = shallow(
      <Radio.Group shape="radius">
        <Radio value="0">选项一</Radio>
        <Radio value="1">选项二</Radio>
        <Radio value="2">选项三</Radio>
      </Radio.Group>
    );
    expect(wrapper.hasClass('shape-radius')).toBe(true);
  });

  // 椭圆角
  it('shape is round', () => {
    const wrapper = shallow(
      <Radio.Group shape="round">
        <Radio value="0">选项一</Radio>
        <Radio value="1">选项二</Radio>
        <Radio value="2">选项三</Radio>
      </Radio.Group>
    );
    expect(wrapper.hasClass('shape-round')).toBe(true);
  });

  // 块级样式
  it('block', () => {
    const wrapper = shallow(
      <Radio.Group block>
        <Radio value="0">选项一</Radio>
        <Radio value="1">选项二</Radio>
        <Radio value="2">选项三</Radio>
      </Radio.Group>
    );
    expect(wrapper.hasClass('block')).toBe(true);
  });

  // 列表样式
  it('type is cell', () => {
    const wrapper = shallow(
      <Radio.Group type="cell">
        <Radio value="0">选项一</Radio>
        <Radio value="1">选项二</Radio>
        <Radio value="2" disabled>选项三</Radio>
      </Radio.Group>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
    expect(wrapper.props().children[0].props.type).toEqual('cell');
  });
});
