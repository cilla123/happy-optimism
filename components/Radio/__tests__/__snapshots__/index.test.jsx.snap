// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio defaultChecked 1`] = `
<div
  class="za-radio theme-primary checked"
>
  <div
    class="za-radio-wrapper"
  >
    <span
      class="za-radio-inner"
    />
    <span
      class="za-radio-text"
    >
      选项一
    </span>
    <input
      checked=""
      class="za-radio-input"
      type="radio"
    />
  </div>
</div>
`;

exports[`Radio renders correctly 1`] = `
<div
  class="za-radio theme-primary"
>
  <div
    class="za-radio-wrapper"
  >
    <span
      class="za-radio-inner"
    />
    <span
      class="za-radio-text"
    >
      选项一
    </span>
    <input
      class="za-radio-input"
      type="radio"
    />
  </div>
</div>
`;

exports[`Radio type is button 1`] = `
<a
  aria-disabled="false"
  class="za-button za-radio theme-primary theme-primary size-sm bordered"
  role="button"
>
  <span>
    <input
      class="za-radio-input"
      type="radio"
      value="0"
    />
    选项一
  </span>
</a>
`;

exports[`Radio type is cell 1`] = `
<div
  class="za-cell is-link"
>
  <div
    class="za-cell-inner"
  >
    <div
      class="za-cell-header"
    />
    <div
      class="za-cell-body"
    >
      <div
        class="za-cell-content"
      >
        <input
          class="za-radio-input"
          type="radio"
          value="0"
        />
        选项一
      </div>
    </div>
    <div
      class="za-cell-footer"
    />
  </div>
</div>
`;

exports[`Radio.Group Radio checked 1`] = `
<div
  class="za-radio-group"
>
  <div
    class="za-radio theme-primary"
  >
    <div
      class="za-radio-wrapper"
    >
      <span
        class="za-radio-inner"
      />
      <span
        class="za-radio-text"
      >
        选项一
      </span>
      <input
        class="za-radio-input"
        type="radio"
      />
    </div>
  </div>
  <div
    class="za-radio theme-primary checked"
  >
    <div
      class="za-radio-wrapper"
    >
      <span
        class="za-radio-inner"
      />
      <span
        class="za-radio-text"
      >
        选项二
      </span>
      <input
        checked=""
        class="za-radio-input"
        type="radio"
      />
    </div>
  </div>
  <div
    class="za-radio theme-primary"
  >
    <div
      class="za-radio-wrapper"
    >
      <span
        class="za-radio-inner"
      />
      <span
        class="za-radio-text"
      >
        选项三
      </span>
      <input
        class="za-radio-input"
        type="radio"
      />
    </div>
  </div>
</div>
`;

exports[`Radio.Group defaultValue 1`] = `
<div
  class="za-radio-group"
>
  <div
    class="za-radio theme-primary"
  >
    <div
      class="za-radio-wrapper"
    >
      <span
        class="za-radio-inner"
      />
      <span
        class="za-radio-text"
      >
        选项一
      </span>
      <input
        class="za-radio-input"
        type="radio"
      />
    </div>
  </div>
  <div
    class="za-radio theme-primary checked"
  >
    <div
      class="za-radio-wrapper"
    >
      <span
        class="za-radio-inner"
      />
      <span
        class="za-radio-text"
      >
        选项二
      </span>
      <input
        checked=""
        class="za-radio-input"
        type="radio"
      />
    </div>
  </div>
  <div
    class="za-radio theme-primary"
  >
    <div
      class="za-radio-wrapper"
    >
      <span
        class="za-radio-inner"
      />
      <span
        class="za-radio-text"
      >
        选项三
      </span>
      <input
        class="za-radio-input"
        type="radio"
      />
    </div>
  </div>
</div>
`;

exports[`Radio.Group renders correctly 1`] = `
<div
  class="za-radio-group"
>
  <div
    class="za-radio theme-primary checked"
  >
    <div
      class="za-radio-wrapper"
    >
      <span
        class="za-radio-inner"
      />
      <span
        class="za-radio-text"
      >
        选项一
      </span>
      <input
        checked=""
        class="za-radio-input"
        type="radio"
      />
    </div>
  </div>
  <div
    class="za-radio theme-primary"
  >
    <div
      class="za-radio-wrapper"
    >
      <span
        class="za-radio-inner"
      />
      <span
        class="za-radio-text"
      >
        选项二
      </span>
      <input
        class="za-radio-input"
        type="radio"
      />
    </div>
  </div>
  <div
    class="za-radio theme-primary"
  >
    <div
      class="za-radio-wrapper"
    >
      <span
        class="za-radio-inner"
      />
      <span
        class="za-radio-text"
      >
        选项三
      </span>
      <input
        class="za-radio-input"
        type="radio"
      />
    </div>
  </div>
</div>
`;

exports[`Radio.Group type is cell 1`] = `
<div
  className="za-radio-group"
>
  <Radio
    block={false}
    checked={true}
    disabled={false}
    key="0/.0"
    onChange={[Function]}
    prefixCls="za-radio"
    theme="primary"
    type="cell"
    value="0"
  >
    选项一
  </Radio>
  <Radio
    block={false}
    checked={false}
    disabled={false}
    key="1/.1"
    onChange={[Function]}
    prefixCls="za-radio"
    theme="primary"
    type="cell"
    value="1"
  >
    选项二
  </Radio>
  <Radio
    block={false}
    checked={false}
    disabled={true}
    key="2/.2"
    onChange={[Function]}
    prefixCls="za-radio"
    theme="primary"
    type="cell"
    value="2"
  >
    选项三
  </Radio>
</div>
`;
